### ISSUE TITLE
Proxy support

### ISSUE THREAD

**Issue by tomchristie** [encode/httpx#36]
Best place to look would be `urllib3`.
Will need to be handled by the "dispatch" layer.

**Comment by sethmlarson** [comment]
Thoughts on implementing sans-I/O proxy handlers as a part of connection generation? Most proxy interchanges aren't as complex.

**Comment by tomchristie** [comment]
I’ve not really looked into what needs to change for proxy support on each of http, https.

Is there a good resource to point at, or summary of the difference in how the request is formed in each of those two cases?

**Comment by sethmlarson** [comment]
HTTP/HTTPS: Send a request with method=CONNECT, target=[host]:[port] along with proxy headers, if you get a 2XX response back then you start using that TCP connection as a tunnel immediately.
SOCKS4: Similar but uses it's own protocol to request a TCP tunnel. Has own authentication.
SOCKS4a: Same as above but DNS resolution can occur on the client side or proxy-side.
SOCKS5: Same as above but different protocol.

Basically all proxies have three states: Waiting for Response, Proxy Tunnel Ready, and Proxy Error.

**Comment by GnubiBORED** [comment]
is http/s proxy support for httpx currently?

**Comment by sethmlarson** [comment]
Not currently, we would happily accept a PR from anyone willing to tackle even one of the above proxy scenarios. :)

**Comment by florimondmanca** [comment]
@sethmlarson Closing this since I think we now have the basics in (HTTP/HTTPS, forward/tunnel) and SOCKS proxy is tracked in #203. If I missed something anyone please reopen. :)
